barretenberg/cpp/src/barretenberg/commitment_schemes/commitment_key.hpp

```diff
@@ -81,6 +82,65 @@ template <class Curve> class CommitmentKey {
         return scalar_multiplication::pippenger_unsafe<Curve>(
             const_cast<Fr*>(polynomial.data()), srs->get_monomial_points(), degree, pippenger_runtime_state);
     };
+
+    /**
+     * @brief Efficiently commit to a sparse polynomial
+     * @details Iterate through the {point, scalar} pairs that define the inputs to the commitment MSM, maintain (copy)
+     * only those for which the scalar is nonzero, then perform the MSM on the reduced inputs.
+     * @warning Method makes a copy of all {point, scalar} pairs that comprise the reduced input. Will not be efficient
+     * in terms of memory or computation for polynomials beyond a certain sparseness threshold.
+     *
+     * @param polynomial
+     * @return Commitment
+     */
+    Commitment commit_sparse(std::span<const Fr> polynomial)
+    {
+        BB_OP_COUNT_TIME();
+        const size_t degree = polynomial.size();
+        ASSERT(degree <= srs->get_monomial_size());
+
+        // Extract the precomputed point table (contains raw SRS points at even indices and the corresponding
+        // endomorphism point (\beta*x, -y) at odd indices).
+        G1* point_table = srs->get_monomial_points();
+
+        // Define structures needed to multithread the extraction of non-zero inputs
+        const size_t num_threads = degree >= get_num_cpus_pow2() ? get_num_cpus_pow2() : 1;
+        const size_t block_size = degree / num_threads;
+        std::vector<std::vector<Fr>> thread_scalars(num_threads);
```

Comment:
My above comment also means that you might get fewer real threads than the number of elements. IIUC, this means that your division of vectors here is still thread-safe, but less efficient than you might think (i.e., not one "bucket" per thread).